exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/ComplexUnnestVisitor.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    Correlate newCorrelate = correlate.copy(correlate.getTraitSet(),
+        leftInputs.get(correlate.getCorrelationId()), right,
+        updatedCorrelationIds.get(correlate.getCorrelationId()),
+        ImmutableBitSet.of(left.getRowType().getFieldCount()), correlate.getJoinType());
+
+    RelBuilder builder = DrillRelFactories.LOGICAL_BUILDER.create(correlate.getCluster(), null);
+    builder.push(newCorrelate);
+
+    List<RexNode> topProjectExpressions = left.getRowType().getFieldList().stream()
+        .map(field -> builder.getRexBuilder().makeInputRef(left, field.getIndex()))
+        .collect(Collectors.toList());
+
+    switch (correlate.getJoinType()) {
+      case LEFT:
+      case INNER:
+        // adds field from the right input of correlate to the top project
+        topProjectExpressions.add(
+            builder.getRexBuilder().makeInputRef(newCorrelate, topProjectExpressions.size() + 1));
+        // fall through
+      case ANTI:
+      case SEMI:
+        builder.project(topProjectExpressions, correlate.getRowType().getFieldNames());
+    }
+    return builder.build();
+  }
+
+  @Override
+  public RelNode visit(RelNode other) {
+    if (other instanceof Uncollect) {
+      return visit((Uncollect) other);
+    }
+    return super.visit(other);
+  }
+
+  public RelNode visit(Uncollect uncollect) {
+    RelBuilder builder = DrillRelFactories.LOGICAL_BUILDER.create(uncollect.getCluster(), null);
+    RexBuilder rexBuilder = builder.getRexBuilder();
+
+    assert uncollect.getInput() instanceof Project : "Uncollect should have Project input";
+
+    Project project = (Project) uncollect.getInput();
+    // If project below uncollect contains only field references, no need to rewrite it
+    if (project.getChildExps().iterator().next().getKind() == SqlKind.FIELD_ACCESS) {
+      return uncollect;
+    }
+
+    // Collects CorrelationId instances used in current rel node
+    RelOptUtil.VariableUsedVisitor variableUsedVisitor = new RelOptUtil.VariableUsedVisitor(null);
+    project.accept(variableUsedVisitor);
+
+    assert variableUsedVisitor.variables.size() == 1 : "Uncollect supports only single correlated reference";
+
+    CorrelationId oldCorrId = variableUsedVisitor.variables.iterator().next();
+    RelNode left = leftInputs.get(oldCorrId);
+
+    // Creates new project to be placed on top of the left input of correlate
+    List<RexNode> leftProjExprs = new ArrayList<>();
+
+    List<String> fieldNames = new ArrayList<>();
+    for (RelDataTypeField field : left.getRowType().getFieldList()) {
+      leftProjExprs.add(rexBuilder.makeInputRef(left, field.getIndex()));
+      fieldNames.add(field.getName());
+    }
+    fieldNames.add(COMPLEX_FIELD_NAME);
+
+    builder.push(left);
+
+    // Adds complex expression with replaced correlation
+    // to the projected list from the left
+    leftProjExprs.addAll(new RexFieldAccessReplacer(builder).apply(project.getChildExps()));
```

Comment:
Assuming that only one project expression is present at this point, Is it better to have an assert here to make sure that only one proj expression is added?

Reply:
Agree with you. I remember some of checks at the validation stage, but one more wouldn't be excessive.
Added in the code above.